Update Interoperability LIPs (#148)
LiskHQ/lips: proposals/lip-0045.md

> To achieve interoperability, chains in the Lisk ecosystem (the mainchain and the sidechains participating in interoperability) exchange information via _cross-chain transactions_. As an example, a user can issue a cross-chain transaction to transfer part of its mainchain LSK balance to a sidechain, and subsequently use it to issue one of the sidechain custom commands.

Reviewer: ```suggestion
To achieve interoperability, chains in the Lisk ecosystem (the mainchain and the sidechains participating in interoperability) exchange information via _cross-chain transactions_. As an example, a user can issue a cross-chain transaction to transfer part of their mainchain LSK balance to a sidechain, and subsequently use it to issue one of the sidechain commands.
```

Reviewer: I think we do not want to use the term custom as in custom transaction or similarly custom commands any more as conceptually all commands are the on the same level.

Author: updated

In the Lisk protocol, cross-chain transactions are special in that, after being included in their origin ledger, some new data structures, the _cross-chain messages_ (CCMs), are created, which are transmitted to other chains via special commands called _cross-chain update_ (CCU) commands. CCUs also contain information about the chain, in the form of a _certificate_. 

> In the receiving chain, CCMs trigger the execution of _cross-chain commands_, in a way similar to how transaction trigger the execution of regular commands.

Author: I think a more complete explanation is also given in the new chain architecture LIP by Shu

One of the main motivations for proposing a [certificate-based communication](https://lisk.com/blog/research/introduction-blockchain-interoperability) between chains is the ability for all chains in the ecosystem to remain state machines. This means that all information needed to validate and advance a chain to the next block is present on the chain itself.  In the Lisk interoperability solution, the only information that needs social consensus is information about the initial validators of the other chain. On the mainchain, this information is received via the sidechain registration command; on sidechains, in a similar way, via the mainchain registration command.

> 

Reviewer: I cannot directly comment on the line below, but you write

> The name and network ID of the mainchain are global protocol constants in the whole ecosystem.

Here the chain ID of the Lisk Mainchain is also a global protocol constant, maybe even more important than the name. Chain names in principle would not need to be stored on the sidechains.

Author: mh, can you suggest an alternative? I find it clear to be honest

Reviewer: I think the current formulation is also fine, I resolve this comment.

### Liveness Condition

Active sidechains are required to prove their liveness to the mainchain at least once every 30 days. 
This is done by including a CCU on the mainchain. 
If a sidechain does not post a CCU within 30 days, the sidechain account is terminated. 
This rule guarantees that users do not send funds to inactive sidechains and that users who have tokens in a sidechain which stops communicating with the ecosystem can recover their tokens.

> There are extra requirements for the first CCU posted on the mainchain: It must contain a non-empty `certificate` which follows the schema defined in [LIP 0061][lip-0061#certificate-schema]; furthermore, if this first CCU also contains a non-empty `inboxUpdate`, the certificate is only valid if it allows the sidechain account to remain live for a reasonable amount of time.

Author: I would try to limit LIP changes to protocol and try not to do cosmetic changes that are not completely necessary

#### Storage of the Outbox Root

> The outbox root property is duplicated and additionally stored separately from all other properties of the chain account. The substore prefix of the outbox root is the constant `SUBSTORE_PREFIX_OUTBOX_ROOT` and the store key is the hash of the chain ID of the partner chain. Storing the outbox root with a different substore prefix allows to separate the subtree corresponding to the outbox roots from the rest of the data in the interoperability store. This choice allows for shorter inclusion proofs for the outbox root, as the other properties of the Interoperability module are not needed to recalculate the state root if the outbox root is known. In particular, the inclusion proof contained in a CCU from a sidechain posted on the mainchain will contain only one hash.

Author: - fixed
- added
- fixed

| `MIN_MODULE_NAME_LENGTH`                      | uint32 | 1                    | The minimum length of a string specifying the name of a module.  |
| `MAX_MODULE_NAME_LENGTH`                      | uint32 | 32                   | The maximum length of a string specifying the name of a module.  |
| `MIN_CROSS_CHAIN_COMMAND_NAME_LENGTH`         | uint32 | 1                    | The minimum length of a string specifying the name of a command. |
| `MAX_CROSS_CHAIN_COMMAND_NAME_LENGTH`         | uint32 | 32                   | The maximum length of a string specifying the name of a command. |
| `MIN_CHAIN_NAME_LENGTH`                       | uint32 | 1                    | The minimum length of a string specifying the name of a chain.   |
| `MAX_CHAIN_NAME_LENGTH`                       | uint32 | 32                   | The maximum length of a string specifying the name of a chain.   |

### Type Definition

| Name               | Type    | Validation            | Description                     |
|--------------------|---------|-----------------------|---------------------------------|
| `Address`          | bytes   | Must be of length `ADDRESS_LENGTH`. | Address of an account.   |

> | `ModuleID`         | bytes   | Must be of length `MODULE_ID_LENGTH`. | ID of a module.   |

Author: thanks, fixed

| `CrossChainCommandID` | bytes   | Must be of length `CROSS_CHAIN_COMMAND_LENGTH`. | ID of a cross-chain command.   |
| `ChainID`          | bytes   | Must be of length `CHAIN_ID_LENGTH`. | ID of a chain.   |

> | `CCU`              | object   | The parameters must follow the `crossChainUpdateTransactionParams` schema. | The type of cross-chain update transactions.   |

Author: We reintroduced the `CCU` type

> | `CCU`              | object   | The parameters must follow the `crossChainUpdateTransactionParams` schema. | The type of cross-chain update transactions.   |

Reviewer 2: We agreed to not include it, so I guess should be removed.

> | `CCM`              | object   | Must follow the `crossChainMessageSchema` schema. | The type of cross-chain messages.   |

Author: done

| Name                                          | Type   | Value                | Description                                                      |
|-----------------------------------------------|--------|----------------------|------------------------------------------------------------------|
| **Interoperability Constants**                |        |                      |                                                                  |
| `MODULE_ID_INTEROPERABILITY`                  | bytes  | TBD                  | ID of the Interoperability module.                               |

> | `OWN_CHAIN_ID`                                | bytes  | 0x00000000           | Local chain ID of the chain.                                     |

Reviewer: As far as I see, this constant is not used for anything in the interoperability module. I also don't understand what a "local chain ID" is.

Author: indeed this concept is now unnecessary. It was used by the token module

> | `MAINCHAIN_ID`                                | bytes  | 0x00000001           | Chain ID of the Lisk mainchain.                                  |

Reviewer: I think it would be nice to follow the same pattern here as for other constants, by starting with the type of constant and then giving the specific name, i.e., here `CHAIN_ID_MAINCHAIN`. Similarly, below `CHAIN_NAME_MAINCHAIN` or `NETOWRK_ID_MAINCHAIN`.

Author: updated

| `CCM_PROCESSED_STATUS_FAILED`                 | uint32 | 6                    | Status code for CCM processed event: failed.                     |
| `CCM_PROCESSED_STATUS_DISCARDED`              | uint32 | 7                    | Status code for CCM processed event: discarded.                  |
| `CCM_PROCESSED_STATUS_RECOVERY_FORWARDED`     | uint32 | 8                    | Status code for CCM processed event: recovered message forwarded.|
| `CCM_PROCESSED_STATUS_RECOVERY_APPLIED`       | uint32 | 9                    | Status code for CCM processed event: recovered message applied.  |
| `CCM_PROCESSED_STATUS_RECOVERY_FAILED`        | uint32 | 10                   | Status code for CCM processed event: recovered message failed.   |
| `CCM_PROCESSED_STATUS_RECOVERY_DISCARDED`     | uint32 | 11                   | Status code for CCM processed event: recovered message discarded.|
| **General Constants**                         |        |                      |                                                                  |
| `EMPTY_HASH`                                  | bytes  | SHA-256("")          | Hash of empty bytes.                                             |

> | `LIVENESS_LIMIT`                              | uint32 | 30x24x3600           | The maximum time interval for the liveness condition.            |

Reviewer: ```suggestion
| `LIVENESS_LIMIT`                              | uint32 | 30 * 24 * 3600           | The maximum time interval for the liveness condition.            |
```

| `Address` | bytes   | Must be of length `ADDRESS_LENGTH`. | Address of an account.   |
| `ModuleID` | bytes   | Must be of length `MODULE_ID_LENGTH`. | ID of a module.   |
| `CrossChainCommandID` | bytes   | Must be of length `CROSS_CHAIN_COMMAND_ID_LENGTH`. | ID of a cross-chain command.   |
| `ChainID` | bytes   | Must be of length `CHAIN_ID_LENGTH`. | ID of a chain.   |
| `CCM` | object   | Must follow the `crossChainMessageSchema` schema (defined in [LIP 0049][lip-0049#ccmschema]). | The type of cross-chain messages.   |
| `CCMID` | bytes   | Must be of length `HASH_LENGTH`. | ID of a cross-chain message.   |
| `MerkleRoot` | bytes   | Must be of length `HASH_LENGTH`. | The root of a Merkle tree.   |
| `PublicKeyEd25519` | bytes | Must be of length `ED25519_PUBLIC_KEY_LENGTH`. | Used for Ed25519 public keys. |
| `PublicKeyBLS` | bytes | Must be of length `BLS_PUBLIC_KEY_LENGTH`. | Used for BLS keys. |
| `SignatureBLS` | bytes | Must be of length `BLS_SIGNATURE_LENGTH`. | Used for BLS signatures. |
| `ActiveValidator` | object | Must have `blsKey: PublicKeyBLS` and `bftWeight: uint64` properties. | The BLS key and BFT weight of an active validator. |

> ### Interoperability Module Store

Reviewer: I don't complete understand the distinction between chain and channel, and why three substores are used to store chain-related information under the key `chainID`.

Author: Originally all properties were part of a single chain account. Then we decided to split them into those relative to the chain account and those relative to the channel. The outbox root is stored separately only to have shorter inclusion proofs. 

The idea is that chain and channel accounts are updated by seprate properties of the ccu and correspond to different concepts.

Reviewer: Thanks for the clarification.

> ##### Store Prefix, Store Key, and Store Value

Reviewer: > The outbox root substore holds the root of the outbox of each partner chain.

I find this description misleading. It is not the outbox root of each partner chain, but rather the outbox root with respect to a partner chain, i.e., the root of the messages target that chain.

Reviewer: This point is still not addressed.

Author: edited

Reviewer: Looks good.

```java
chainAccountSchema = {
    "type": "object",

>     "required": ["name", "lastCertificate", "status"],

Author: Added an extra explanation in the text. This is also defined in LIP43.

                },
                "root": {
                    "dataType": "bytes",
                    "length": HASH_LENGTH,
                    "fieldNumber": 3
                }
            }
        },
        "partnerChainOutboxRoot": {
            "dataType": "bytes",

>             "length": HASH_LENGTH,

Reviewer: You write below:

>The value of this property is set to the outbox root computed from the last CCU from the partner chain. 

Isn't it rather "obtained" from the last CCU then "computed" from the last CCU?

Author: you compute it if there are ccms attached in the ccu

Reviewer: I though you only ever copy the `outboxRoot` from the CCU to the `partnerChainOutboxRoot` and only the inbox root is something updated via computation.

Reviewer: Okay, can be resolved.

            "fieldNumber": 3
        },

>         "messageFeeTokenID": {

Reviewer: In the description below it says:

>  This property is an object storing a chainID and a localID. These values together form the token ID of the token used to pay for the cross-chain message fees. The default values are chainID = 1 and localID = 0, corresponding to the LSK token.

I think this is outdated and still from the time when token IDs where objects on not fixed byte length identifiers.

Author: fixed

            "dataType": "bytes",
            "length": TOKEN_ID_LENGTH,
            "fieldNumber": 4
        }
    }

> }

Author: fixed. In the verification of the registration message (LIP49)

* `partnerChainOutboxSize`: This property corresponds to the size of the outbox tree of the partner chain, i.e. the number of cross-chain messages sent from the partner chain. This property is updated by the [`inboxUpdate property`][research:ccu] contained in CCUs from the partner chain. The default value of this property is 0.
* `partnerChainInboxSize`: This property corresponds to the size of the inbox tree of the partner chain, i.e. the number of cross-chain messages received and processed by the partner chain. This property is used to verify the validity of the message recovery command and it is updated by a [cross-chain update receipt message][research:ccm] from the partner chain. The default value of this property is 0.
* `name`: This property corresponds to the name of the sidechain as a string of characters. It has to be unique in the ecosystem. For the mainchain account on a sidechain, this property is initialized to the string `MAINCHAIN_NAME`. For a sidechain account on the mainchain, this property is set by the sidechain registration command.

> * `status`: This property stores the current status of the partner chain account. As explained [above](#Life-Cycle-of-a-Sidechain), there are 3 possible statuses: ''active'', ''registered'', and ''terminated''. The default value of this property is `CHAIN_REGISTERED`, corresponding to the "registered" status.

Author: just removed that sentence

            "fieldNumber": 1
        },

>         "ID": {

Reviewer: I think `chainID` would be more clear. Note that in the description below, you already assume that this is the property name.

    "properties": {
        "stateRoot": {
            "dataType": "bytes",

>             "length": HASH_LENGTH,

Author: yes, fixed

    "required": ["ID"],
    "properties": {

>         "ID": {

Reviewer: I think `chainID` would be more clear.

Reviewer: Note that below you also use a different property name

> The store value is an object with `chainID=MAINCHAIN_ID`, serialized using the JSON schema `chainIDSchema`.

Here I would also recommend to write `chainID == MAINCHAIN_ID`

Author: fixed

    inboxTree = partial Merke tree with properties from channel(chainID).inbox
    inboxTree.append(sha256(appendData))
    channel(chainID).inbox.root = inboxTree.root
    channel(chainID).inbox.appendPath = inboxTree.appendPath
    channel(chainID).inbox.size = inboxTree.size
```

#### appendToOutboxTree

The `appendToOutboxTree` function appends a new element to the underlying Merkle tree of the outbox of a chain account.

```python
def appendToOutboxTree(chainID: ChainID, appendData: bytes) -> None:
    # Update root, appendPath, and size
    outboxTree = partial Merke tree with properties from channel(chainID).outbox
    outboxTree.append(sha256(appendData))
    channel(chainID).outbox.root = outboxTree.root
    channel(chainID).outbox.appendPath = outboxTree.appendPath
    channel(chainID).outbox.size = outboxTree.size
```

#### addToOutbox

The `addToOutbox` function adds a new CCM to the outbox of a chain account.

```python
def addToOutbox(chainID: ChainID, ccm: CCM) -> None:
    ccmBytes = encode(crossChainMessageSchema, ccm)
    appendToOutboxTree(chainID, ccmBytes)
    outboxRoot(chainID) = channel(chainID).outbox.root
```

#### isLive

> The `isLive` function checks the liveness requirement for the partner chain. Notice that this requirement is only applied on the mainchain.

Author: I added an explanatory sentence (which was for sure needed), let me know if it's fine.

    "required": ["ID"],
    "properties": {

>         "ID": {

Reviewer: I would call this property `chainID`. As we handle different IDs here (e.g., also the network ID), this makes things more explicit.

Reviewer: Again the description below is inconsistent with the schema:

> The store value is an object with chainID=MAINCHAIN_ID, serialized using the JSON schema chainIDSchema.

Reviewer: I resolve this point as it is fixed.

> The `appendToInboxTree` function appends a new element to the underlying Merkle tree of the inbox of a chain account.

Author: You mean the functions used to verify the CCU? I think it makes more sense to have them there, since they are used only for CCU processing

> The `appendToInboxTree` function appends a new element to the underlying Merkle tree of the inbox of a chain account.

Author: I see, added

```python
def appendToInboxTree(chainID: ChainID, appendData: bytes) -> None:
    # Update root, appendPath, and size

>     inboxTree = Merkle tree of channel(chainID).inbox

Reviewer: Here I assume the whole Merkle tree is not necessary, rather the append path, root and size is sufficient. Maybe this could be made more clear.

Author: yes. Here we basically instantiate a 'light Merkle tree' from the root, append path, and size (appendix B in LIP31). I think this formulation is fine, but if you have a better suggestion I can change it

Reviewer: Maybe 

>  inboxTree = partial Merke tree with properties from channel(chainID).inbox

or simply 

> inboxTree = channel(chainID).inbox

The phrase "Merkle tree of" sounds like an operation is performed, where really it is just coping the data. What is a bit odd that directly below you assign each property separately whereas here you do it all in one line.

Author: We do perform an operation tho, we have to instatiate a Merkle tree object, and in particular call the method `update` on it. That's why I think `inboxTree = channel(chainID).inbox` is incorrect.

Reviewer: You mean the `append` call we make in the next line? How about the following suggestion then: 

> inboxTree = partial Merke tree with properties from channel(chainID).inbox

Author: ok, fixed

    inboxTree.append(SHA-256(appendData))
    channel(chainID).inbox.root = inboxTree.root
    channel(chainID).inbox.appendPath = inboxTree.appendPath
    channel(chainID).inbox.size = inboxTree.size
```

#### appendToOutboxTree

The `appendToOutboxTree` function appends a new element to the underlying Merkle tree of the outbox of a chain account.

```python
def appendToOutboxTree(chainID: ChainID, appendData: bytes) -> None:
    # Update root, appendPath, and size

>     outboxTree = Merkle tree of channel(chainID).outbox

Reviewer: Here I assume the whole Merkle tree is not necessary, rather the append path, root and size is sufficient. Maybe this could be made more clear.

Reviewer: Similar suggestion as above.

Author: fixed

The `forward` function forwards a CCM to the specified receiving chain. This function is only called on the mainchain when processing a CCU. It is specified in [LIP 0049][lip-0049].

#### bounce and bounceWithoutFee

The `bounce` and `bounceWithoutFee` functions return a CCM that errored during the CCM execution stages to the sending chain. 
`bounceWithoutFee` is called if an error occurs during the cross-chain command execution (see [`apply` function][lip-0049#apply]) in the receiving chain. 
In this case, the fee property is set to 0 and the message is returned only if the fee was larger than a minimum return fee and the status of the CCM is `CCM_OK`.
`bounce` is called if an error occurs in any other stage.
In this case, the fee is not changed and the message is returned if the status of the CCM is `CCM_OK`.

```python

> def bounce(ccm: CCM, errorCode: uint32) -> None:

Reviewer: As raised on Slack, we should check if we should still have both `bounce` and `bounceWithoutFee`.

Author: solved. We only keep bounce.

    ccmID = sha256(encode(crossChainMessageSchema, ccm))
    emitEvent(
        module = MODULE_NAME_INTEROPERABILITY,

>         name = EVENT_NAME_CCM_PROCESSED,

Reviewer: I find this event overall a bit confusing and we also do not use it consistently. Here, for instance, we use it both to emit an event about the kind of error that happened (which at the same time is the status of the returned message) as well as to say what happens further with the CCM. In some cases (`ERROR_CODE_INVALID_FORMAT`), we emit only an event about the error, but do not say what happens with the CCM. I wonder if it would not be sufficient to just emit one event in all cases.

Reviewer: What I find further confusing is that we have constants from 0 (`CCM_OKAY`) to 11 (`ERROR_CODE_MESSAGE_FEE_ERROR`) that refer to very different aspects of the CCM processing. For instance:

- `CCM_OK` is used both as default `status` code when sending a message (basically just meaning, this is not a return error CCM) and to signal that a CCM was executed successfully, which are quite different things.
- `CCM_APPLIED` is used for a CCM processed event in case of an error although it suggests that it means that a CCM is successfully applied?
- We have `ERROR_CODE_INVALID_CCM` and `ERROR_CODE_FAILED_CCM`, which seem rather generic, but then there are more specific errors for the case of invalid and failing CCMs.

Reviewer: Maybe the event could have the properties:

- "executionResult" in {"success", "failed", "invalid"} 
- "action" in {"apply", "bounce", "forward", "discard"}
- "errorCode" in { error codes already defined for different failures}

Of course property and value names could be different, in particular, or values could also be integers.

Note that any message should either be successful, failing or invalid, and the action of the interoperability module should always be either apply (in case the CCM is successful), bounce (this can happen for failing or invalid CCMs), forward or discard.

Author: Regarding the first point: it was a copy/paste mistake, when we apply we use `CCM_APPLIED`. 

I would reserve 'invalid' and 'failed' for the error that you get during execution stack. if the channel is unavailable, it makes sense to use a different error code. 

Regarding your proposal, I still think there are some unclear points: for instace, what would be the executionResult in case of action=forwarding?

Author: addressed together

Reviewer: Looks good now.

        data = {"ccmID": ccmID, "result": errorCode},
        topics = [ccm.sendingChainID, ccm.receivingChainID]
    )

    if ccm.status == CCM_OK:
        ccm.status = errorCode
        ccm.sendingChainID, ccm.receivingChainID = ccm.receivingChainID, ccm.sendingChainID
        addToOutbox(ccm.receivingChainID, ccm)
        newCcmID = sha256(encode(crossChainMessageSchema, ccm))
        emitEvent(
            module = MODULE_NAME_INTEROPERABILITY,
            name = EVENT_NAME_CCM_PROCESSED,

>             data = {"ccmID": newCcmID, "result": CCM_BOUNCED},

Reviewer: What is confusing here is that usually the CCM Processed event has the ID of the CCM being processed as data where instead here it is a new CCM being emitted.

Reviewer: Wouldn't it be better to use the `EVENT_NAME_CCM_SENT` event?

Author: done

#### bounce

The `bounce` function returns a CCM that errored during the CCM execution stages to the sending chain if the message fee is larger than a minimum return fee (that depends on the message length) and the status of the CCM is `CCM_OK`.
If the function is called during the cross-chain command execution (see [`apply` function][lip-0049#apply]), the error code will be `ERROR_CODE_FAILED_CCM`. 
In this case the message fee is set to 0, as it was already assigned to the relayer.
Otherwise, the fee is reduced by the minimum fee.

```python
def bounce(ccm: CCM, errorCode: uint32) -> None:
    ccmID = sha256(encode(crossChainMessageSchema, ccm))

>     emitEvent(

Author: I think we should leave to define what happens to the ccm.

        module = MODULE_NAME_INTEROPERABILITY,
        name = EVENT_NAME_CCM_PROCESSED,
        data = {"ccmID": ccmID, "result": errorCode},
        topics = [ccm.sendingChainID, ccm.receivingChainID]
    )

    minimumFee = MIN_RETURN_FEE * len(encode(crossChainMessageSchema, ccm))
    if ccm.status == CCM_OK and ccm.fee >= minimumFee:
        ccm.status = errorCode
        ccm.sendingChainID, ccm.receivingChainID = ccm.receivingChainID, ccm.sendingChainID
        
        # If the function is called during the cross-chain command execution, the fee is set to 0
        if errorCode == ERROR_CODE_FAILED_CCM:
            ccm.fee = 0
        else:

>             ccm.fee -= minimumFee     

Author: for the moment, I would just burn it for simplicity. In this case, notice that we do not need to do anything. @janhack opinions?

>             ccm.fee -= minimumFee     

Reviewer: I think burning for simplicity is fine as it would only happen in case of chain termination.

>             ccm.fee -= minimumFee     

Author: If we burn, we do not need to modify anything. If you agree, we can resolve this

>             ccm.fee -= minimumFee     

Reviewer 2: I would prefer to send it to the relayer, but for now it might be ok, since we will revisit this soon.

The `forward` function forwards a CCM to the specified receiving chain. This function is only called on the mainchain when processing a CCU. It is specified in [LIP 0049][lip-0049].

#### createTerminatedStateAccount

The `createTerminatedStateAccount` function creates an entry in the terminated state substore.

```python
def createTerminatedStateAccount(chainID: ChainID, stateRoot: bytes = EMPTY_BYTES) -> None:    
    if chainAccount(chainID) exists:
        chainAccount(chainID).status = CHAIN_STATUS_TERMINATED
        remove the entry with storeKey = chainID from the outbox root substore

        # If no stateRoot is given as input, get it from the state
        if len(stateRoot) == 0:
            stateRoot = chainAccount(chainID).lastCertificate.stateRoot

        terminatedState = {
            "stateRoot": stateRoot,
            "mainchainStateRoot": EMPTY_BYTES,
            "initialized": True
        }

        # Emit chain status updated event.

>         emitEvent(

Reviewer: I am not sure it is so nice here to only have an event in this case, but not the case below.

Author: The idea is that we emit the event whenever we update the chain account. On a sidechain there is no chain account to update tho. One option would be to introduce a new event when the terminated state account is created

Reviewer: I think it would be useful to have an event here. Otherwise, it may be difficult to find out at which stage and for what reason a chain is terminated.

Author: added a new event for creating chain terminated account

Reviewer: Looks good, I resolve this comment.

        }

    create an entry in the terminated state substore with
        storeKey = chainID
        storeValue = encode(terminatedStateSchema, terminatedState)
```

#### createTerminatedOutboxAccount

The `createTerminatedOutboxAccount` function creates an entry in the terminated outbox substore.

```python

> def createTerminatedOutboxAccount(chainID: ChainID, 

Reviewer: Here and in other places, could we follow the [PEP8 guidelines on indentation](https://legacy.python.org/dev/peps/pep-0008/#indentation) as mentioned in the pseudocode document. In particular, the following is explicitly discouraged in PEP8:

```
# Further indentation required as indentation is not distinguishable.
def long_function_name(
    var_one, var_two, var_three,
    var_four):
    print(var_one)
```

Author: fixed

> def createTerminatedOutboxAccount(chainID: ChainID, 

Reviewer: As far as I see, this command is only called from one place right now, the "Message Recovery Initialization Command". This command also does not emit any event. Could it be useful to have an associated event for this command or do you think this is not necessary?

Author: Same as above, here I was thinking that the events emitted when processing that command would be enough, but we could add a new event for this

Reviewer: There is not event emitted by the Message Recovery Initialization Command so you mean just the default event that says the command was executed successfully. This would not provide any information about the state change, but I guess the interoperability module state after the execution would contain the necessary information.

Author: I added two new events for the creation of a terminated state and outbox account respectively.

Reviewer: Looks good now, I resolve this comment.

##### Topics

- `sendingChainID`: The ID of the sending chain.
- `receivingChainID`: The ID of the receiving chain.

##### Data

```java
ccmProcessedDataSchema = {
    "type": "object",
    "required": ["ccmID", "status"],
    "properties": {
        "ccmID": {
            "dataType": "bytes",
            "length": HASH_LENGTH, 
            "fieldNumber": 1
        },
        "status": {
            "dataType": "uint32",
            "fieldNumber": 2
        }
    }
}
```

### Protocol Logic for Other Modules

#### send

The `send` function is used to create and add a message to the outbox of a partner chain.

```python
def send(sendingAddress: Address, 
    module: Module, 
    crossChainCommand: CrossChainCommand, 
    receivingChainID: ChainID, 
    fee: uint64, 
    status: uint32, 
    params: object
) -> None:
    # Processing on the mainchain
    if ownChainAccount.chainID == CHAIN_ID_MAINCHAIN:
        # Chain has to exist on mainchain
        if chainAccount(receivingChainID) does not exists:
            emitPersistentEvent(
                module = MODULE_NAME_INTEROPERABILITY,
                name = EVENT_NAME_CCM_SENT,

>                 data = {"ccmID": ccmID, "status": CCM_PROCESSED_STATUS_CHANNEL_UNAVAILABLE},

Reviewer: ```suggestion
                data = {"ccmID": ccmID, "status": CCM_SENT_STATUS_CHANNEL_UNAVAILABLE},
```

Author: fixed

>                 topics = [ccm.sendingChainID, ccm.receivingChainID, ccmID]

Author: yes, refactored

            )
            raise Exception('Receiving chain does not exist.')
        partnerChainID = receivingChainID
    # Processing on a sidechain
    else:
        # Check for direct channel
        if chainAccount(receivingChainID) does not exists:
            partnerChainID = mainchainID
        else:
            partnerChainID = receivingChainID

    # Chain must be live; This checks is always on the receivingChainID
    if not isLive(receivingChainID):
        emitPersistentEvent(
            module = MODULE_NAME_INTEROPERABILITY,
            name = EVENT_NAME_CCM_SENT,

>             data = {"ccmID": ccmID, "status": CCM_SENT_STATUS_CHANNEL_UNAVAILABLE},

Reviewer: Do you not want to distinguish the cases if the receiving chain is not live (here) or not active (below). Maybe this could allow the front-end to be more specific why sending a CCM failed.

Author: I think this is fine-grained enough

            data = {"ccmID": ccmID, "status": CCM_SENT_STATUS_CHANNEL_UNAVAILABLE},
            topics = [ccm.sendingChainID, ccm.receivingChainID, ccmID]
        )
        raise Exception('Receiving chain is not active.')

    # Create cross-chain message
    paramsSchema = JSON schema associated with (module, crossChainCommand)
    ccm = {
        "nonce": ownChainAccount.nonce,
        "module": module,
        "crossChainCommand": crossChainCommand,
        "sendingChainID": ownChainAccount.chainID,
        "receivingChainID": receivingChainID,
        "fee": fee,
        "status": status,
        "params": encode(paramsSchema, params)
    }

    # Validate ccm size
    try:
        validateFormat(ccm)
    except Exception as e:
        emitPersistentEvent(
            module = MODULE_NAME_INTEROPERABILITY,
            name = EVENT_NAME_CCM_SENT,
            data = {"ccmID": ccmID, "status": CCM_SENT_STATUS_INVALID_FORMAT},
            topics = [ccm.sendingChainID, ccm.receivingChainID, ccmID]
        )
        raise e

    # Pay message fee
    if fee > 0:
        try:
            Token.payMessageFee(sendingAddress, fee)
        except Exception as e:

>             emitPersistentEvent(

Author: I think we should explicitly emit an event for whatever happens to a ccm. If we do not emit here, we do not have an event to signal that the ccm failed

>             emitPersistentEvent(

Author: I think we can emit 2 events in this case, but otherwise I would also remove it from the token module.

>             emitPersistentEvent(

Reviewer: Let us remove the one from the token module then.

>             emitPersistentEvent(

Reviewer 2: Yes, will take care of that. Resolving this comment.

                module = MODULE_NAME_INTEROPERABILITY,
                name = EVENT_NAME_CCM_SENT,
                data = {"ccmID": ccmID, "status": CCM_SENT_STATUS_MESSAGE_FEE_ERROR},
                topics = [ccm.sendingChainID, ccm.receivingChainID, ccmID]
            )
            raise e

    addToOutbox(partnerChainID, ccm)
    ownChainAccount.nonce += 1

    # Emit CCM processed event.

>     ccmID = sha256(encode(crossChainMessageSchema, ccm))

Reviewer: You only compute the CCM ID here, but use it in all events before. Also it may not make sense/be possible to compute the ID if the format validation fails, so maybe the CCM ID needs to be empty there. If you want to use the CCM ID in events, consider doing the format validation in the send function first.

Author: yes, there were several small imprecisions here. Please take a look now.

* [State recovery initialization command][lip-0054]
* [Message recovery initialization command][lip-0054]

> ### Events

Author: I added a small list

##### Topics

- `chainID`: The ID of the chain.

##### Data

The `data` property of this event uses the `chainAccountSchema` schema given [above](#json-schema-1).

#### CrossChainMessageProcessed

This event has `typeID = TYPE_ID_CCM_PROCESSED`. This event is emitted when a CCM is sent or processed.
The status property of the data schema can take one of the following values:

1. `CCM_PROCESSED_STATUS_SENT`: The CCM was sent to the receiving chain.
2. `CCM_PROCESSED_STATUS_CHANNEL_UNAVAILABLE`: The receiving chain is not available.
3. `CCM_PROCESSED_STATUS_MODULE_NOT_SUPPORTED`: The receiving chain does not support the module.
4. `CCM_PROCESSED_STATUS_COMMAND_NOT_SUPPORTED`: The receiving chain does not support the command.
5. `CCM_PROCESSED_STATUS_FORWARDED`: The CCM was forwarded to the receiving chain.
6. `CCM_PROCESSED_STATUS_APPLIED`: The CCM was applied on the receiving chain.
7. `CCM_PROCESSED_STATUS_DISCARDED`: The CCM was discarded.
8. `CCM_PROCESSED_STATUS_RECOVERY_FORWARDED`: The CCM was forwarded to the original sending chain for recovery.
9. `CCM_PROCESSED_STATUS_RECOVERY_APPLIED`: The CCM was recovered and applied on the original sending chain.

> 10. `CCM_PROCESSED_STATUS_RECOVERY_DISCARDED`: The CCM was ricovered and discarded.

Author: We will define constants only in lip45

##### Topics

- `sendingChainID`: The ID of the sending chain.
- `receivingChainID`: The ID of the receiving chain.
- `ccmID`: The ID of the CCM that was sent. This topic is present only for newly sent messages (event status equals `CCM_PROCESSED_STATUS_SENT`).

> Furthermore, one

Author: to be removed :)

##### Data

```java
ccmProcessedDataSchema = {
    "type": "object",
    "required": ["ccmID", "status"],
    "properties": {
        "ccmID": {
            "dataType": "bytes",
            "length": HASH_LENGTH, 
            "fieldNumber": 1
        },
        "status": {
            "dataType": "uint32",
            "fieldNumber": 2
        }
    }
}
```

### Protocol Logic for Other Modules

#### send

The `send` function is used to create and add a message to the outbox of a partner chain.

```python
def send(sendingAddress: Address, 
    moduleID: ModuleID, 
    crossChainCommandID: CrossChainCommandID, 
    receivingChainID: ChainID, 
    fee: uint64, 
    status: uint32, 
    params: object) -> None:
    # Processing on the mainchain
    if ownChainAccount.ID == MAINCHAIN_ID:
        # Chain has to exist on mainchain
        if chainAccount(receivingChainID) does not exists:

>             raise Exception('Receiving chain does not exist.')

Reviewer: Why did you decide not to have an error event in this case?

Author: Makes sense to emit event actually. I moved the ccm sent event to a new event and added errors.

        partnerChainID = receivingChainID
    # Processing on a sidechain
    else:
        # Check for direct channel
        if chainAccount(receivingChainID) does not exists:
            partnerChainID = mainchainID
        else:
            partnerChainID = receivingChainID

    # Chain must be live; This checks is always on the receivingChainID
    if not isLive(receivingChainID):
        raise Exception('Receiving chain is not live.')
    # Chain status must be active

>     if not chainAccount(partnerChainID).status == CHAIN_STATUS_ACTIVE:

Reviewer: Should this check maybe come before checking liveness? For instance, for a terminated chain `isLive` already returns false, so doing this check first may result in a more specific exception case.

Author: chainAccount may not exist tho, so it would make for an extra check first

Reviewer: Okay. I also see now that the liveness check is for the receiving chain and being active is checked for the partner chain.

        "moduleID": moduleID,
        "crossChainCommandID": crossChainCommandID,
        "sendingChainID": ownChainAccount.ID,
        "receivingChainID": receivingChainID,
        "fee": fee,
        "status": status,
        "params": encode(paramsSchema, params)
    }

    # Validate ccm size
    validateFormat(ccm)

    # Pay message fee
    Token.payMessageFee(sendingAddress, ccm)

    addToOutbox(partnerChainID, ccm)
    ownChainAccount.nonce += 1

    # Emit CCM processed event.
    ccmID = SHA-256(encode(crossChainMessageSchema, ccm))

    emitEvent(
        moduleID = MODULE_ID_INTEROPERABILITY,
        typeID = TYPE_ID_CCM_PROCESSED,
        data = {"ccmID": ccmID, "status": CCM_PROCESSED_STATUS_SENT},
        topics = [ccm.sendingChainID, ccm.receivingChainID, ccmID]
    )
```

#### terminateChain

The `terminateChain` function terminates a chain account.

```python
def terminateChain(chainID: ChainID) -> None:
    # Chain was already terminated, do nothing
    if terminatedStateAccount(chainID) exists:
        return

    if chainID == ownChainAccount.ID:
        raise Exception('Chain cannot terminate itself.')

    # Processing on a sidechain

>     if ownChainAccount.ID != MAINCHAIN_ID:

Reviewer: This check is not necessary, right? As the Lisk Mainchain can already not terminate itself.

Author: This function is an API tho.

Reviewer: All good, I was just confused.

        # The mainchain cannot be terminated
        if chainID == MAINCHAIN_ID:
            raise Exception('The mainchain cannot be terminated.')

    send(

>         EMPTY_FEE_ADDRESS,

Reviewer: If the type hinting is properly implemented, then this input parameter would not be accepted. This could be solved maybe by a dummy address (all 0), less restrictive type hinting or an optional parameter.

Author: True. I think in typescript they just have less strict type hinting. Probably we should just have an option to send without fee. I find this rather an implementation detail (the fee address is not used unless the fee is non 0)

Reviewer: Okay, let us keep it then.

        MODULE_ID_INTEROPERABILITY,
        CROSS_CHAIN_COMMAND_ID_CHANNEL_TERMINATED,
        chainID,
        0,
        CCM_STATUS_OK,
        EMPTY_BYTES
    )

    return createTerminatedStateAccount(chainID)
```

#### error

The `error` function is used to add an error code to a CCM and then add it to the outbox of a partner chain.

```python

> def error(ccm: CCM, errorStatus: uint32, relatedCCMIDs: list[bytes]) -> None:

Reviewer: `relatedCCMIDs: list[bytes]` does not seem to be used any more

    # Error codes from 0 to MAX_RESERVED_ERROR_STATUS (included) are reserved to the Interoperability module
    if 0 <= errorStatus <= MAX_RESERVED_ERROR_STATUS:
        raise Exception('Invalid error status.')
    
    originalCCMID = SHA-256(encode(crossChainMessageSchema, ccm))
    return send(
        EMPTY_FEE_ADDRESS,
        ccm.moduleID,
        ccm.crossChainCommandID,
        ccm.sendingChainID,
        0,
        errorStatus,
        ccm.params
    )
```

#### getChainAccount

The `getChainAccount` function returns the object stored in the chain data substore with store key equal to the input `chainID: ChainID`.
It returns `chainAccount(chainID)`.

> #### getChannel

Reviewer: You added the channel validators substore for which there is not getter at the moment.

Author: added

The `getChannel` function returns the object stored in the channel data substore with store key equal to the input `chainID: ChainID`.
It returns `channel(chainID)`.

#### getOwnChainAccount

The `getOwnChainAccount` function returns the object stored in the own chain data substore.
It returns `ownChainAccount`.

#### getTerminatedStateAccount

The `getTerminatedStateAccount` function returns the object stored in the terminated state substore with store key equal to the input `chainID: ChainID`.
It returns `terminatedStateAccount(chainID)`.

#### getTerminatedOutboxAccount

The `getTerminatedOutboxAccount` function returns the object stored in the terminated outbox substore with store key equal to the input `chainID: ChainID`.
It returns `terminatedOutboxAccount(chainID)`.

> #### isDirectChannelActive

Reviewer: I am not sure about the purpose of this method. For chains that exists, the status can be "registered", "active" or "terminated". This method would return "active" also for chains for which the function `isLive` would return `False`, which is kind of misleading. Wouldn't it be more useful to get the chain status instead?

Author: I think it will be used by the direct sidechain LIP. We could also remove it for the moment and only add it later.

Reviewer: I would remove this function then. It is also an API (not an internal function).

Author: removed

* [State recovery initialization command][lip-0054]
* [Message recovery initialization command][lip-0054]

### Cross-chain Commands

The Interoperability module contains the following cross-chain commands:

* [Sidechain terminated cross-chain command][lip-0049]
* [Registration cross-chain command][lip-0049]
* [Channel terminated cross-chain command][lip-0049]
  
### Events

#### ChainAccountUpdated

This event has `name = EVENT_NAME_CHAIN_ACCOUNT_UPDATED`. 
This event is emitted when a chain account is updated.

##### Topics

- `chainID`: The ID of the chain.

##### Data

> The `data` property of this event uses the `mainchainRegParams` schema given in [LIP0043][lip-0043].

Reviewer 3: Why does it use `mainchainRegParams` schema? Shouldn't it be `chainAccountSchema`? In the pseudo code above for sidechain registration it uses chainAccount as data.

Author: I switched the two events! fixed

#### TerminatedOutboxCreated

This event has `name = EVENT_NAME_TERMINATED_OUTBOX_CREATED`. This event is emitted when a new entry is created in the terminated outbox substore.

##### Topics

- `chainID`: The ID of the terminated outbox account.

##### Data

The data property follows the `terminatedOutboxAccountSchema` schema given [above](#json-schema-6).

### Protocol Logic for Other Modules

#### send

The `send` function is used to create and add a message to the outbox of a partner chain.

```python
def send(sendingAddress: Address, 
    module: Module, 
    crossChainCommand: CrossChainCommand, 
    receivingChainID: ChainID, 
    fee: uint64, 
    status: uint32, 
    params: object
) -> None:
# Create cross-chain message
    paramsSchema = JSON schema associated with (module, crossChainCommand)
    ccm = {
        "nonce": ownChainAccount.nonce,
        "module": module,
        "crossChainCommand": crossChainCommand,
        "sendingChainID": ownChainAccount.chainID,
        "receivingChainID": receivingChainID,
        "fee": fee,
        "status": status,
        "params": encode(paramsSchema, params)
    }

    # Validate ccm size
    try:
        validateFormat(ccm)

>     except Exception as e:

Author: I would leave it like this for the moment.

        emitPersistentEvent(
            module = MODULE_NAME_INTEROPERABILITY,
            name = EVENT_NAME_CCM_SENT,
            data = {"ccmID": ccmID, "result": ERROR_CODE_INVALID_FORMAT},
            topics = [ownChainAccount.chainID, receivingChainID, EMPTY_BYTES]
        )
        raise e

    # From now on, we can assume that the ccm is valid.
    ccmID = sha256(encode(crossChainMessageSchema, ccm))

    # Processing on the mainchain
    if ownChainAccount.chainID == CHAIN_ID_MAINCHAIN:
        partnerChainID = receivingChainID
    # Processing on a sidechain
    else:
        # Check for direct channel
        if chainAccount(receivingChainID) does not exists:
            partnerChainID = CHAIN_ID_MAINCHAIN
        else:
            partnerChainID = receivingChainID

    # receivingChainID must correspond to a live chain.
    if not isLive(receivingChainID):
        emitPersistentEvent(
            module = MODULE_NAME_INTEROPERABILITY,
            name = EVENT_NAME_CCM_SENT,
            data = {"ccmID": ccmID, "result": ERROR_CODE_CHANNEL_UNAVAILABLE},
            topics = [ccm.sendingChainID, ccm.receivingChainID, ccmID]
        )

>         raise Exception('Receiving chain is not live.')

Author: it would not get automatically terminated, but a initialize terminate state command can still be sent to trigger the termination

>         raise Exception('Receiving chain is not live.')

Author: This is an open point still, basically: should we force termination for chains violating liveness requirement on the mainchain? @janhack @gkoumout

>         raise Exception('Receiving chain is not live.')

Author: we decided not to, can be resolved from my side

>         raise Exception('Receiving chain is not live.')

Reviewer: Note that all state changes are reverted here as we raise an exception. That is why we decided not to terminate the chain here. Can be resolved from my side.

##### Topics

- `chainID`: The ID of the terminated outbox account.

##### Data

The data property follows the `terminatedOutboxAccountSchema` schema given [above](#json-schema-6).

### Protocol Logic for Other Modules

#### send

The `send` function is used to create and add a message to the outbox of a partner chain.

```python
def send(sendingAddress: Address, 
    module: Module, 
    crossChainCommand: CrossChainCommand, 
    receivingChainID: ChainID, 
    fee: uint64, 
    status: uint32, 
    params: object
) -> None:
# Create cross-chain message
    paramsSchema = JSON schema associated with (module, crossChainCommand)
    ccm = {
        "nonce": ownChainAccount.nonce,
        "module": module,
        "crossChainCommand": crossChainCommand,
        "sendingChainID": ownChainAccount.chainID,
        "receivingChainID": receivingChainID,
        "fee": fee,
        "status": status,
        "params": encode(paramsSchema, params)
    }

    # Validate ccm size
    try:
        validateFormat(ccm)
    except Exception as e:
        emitPersistentEvent(
            module = MODULE_NAME_INTEROPERABILITY,
            name = EVENT_NAME_CCM_SENT,

>             data = {"ccmID": ccmID, "result": ERROR_CODE_INVALID_FORMAT},

Reviewer: `ccmID` is only computed below. I am not sure you can even compute a CCM ID if the format is invalid.

Reviewer: Checking the `validateFormat` function, the encoding should always work and actually only the message might be too large or the module/command name may have invalid format.

Author: fixed

            topics = [ownChainAccount.chainID, receivingChainID, EMPTY_BYTES]
        )
        raise e

    # From now on, we can assume that the ccm is valid.
    ccmID = sha256(encode(crossChainMessageSchema, ccm))

>     # Processing on the mainchain

Reviewer: I would not call this processing, but maybe "sending" or something similar. As the processing is only done in the receiving chain.

Author: all good

#### TerminatedOutboxCreated

This event has `name = EVENT_NAME_TERMINATED_OUTBOX_CREATED`. This event is emitted when a new entry is created in the terminated outbox substore.

##### Topics

- `chainID`: The ID of the terminated outbox account.

##### Data

The data property follows the `terminatedOutboxAccountSchema` schema given [above](#json-schema-6).

### Protocol Logic for Other Modules

#### send

The `send` function is used to create and add a message to the outbox of a partner chain.

```python
def send(sendingAddress: Address,
    module: Module,
    crossChainCommand: CrossChainCommand,
    receivingChainID: ChainID,
    fee: uint64,
    status: uint32,
    params: object
) -> None:

    paramsSchema = JSON schema associated with (module, crossChainCommand)
    # This could fail
    params = encode(paramsSchema, params)

    ccm = {
        "nonce": ownChainAccount.nonce,
        "module": module,
        "crossChainCommand": crossChainCommand,
        "sendingChainID": ownChainAccount.chainID,
        "receivingChainID": receivingChainID,
        "fee": fee,
        "status": status,
        "params": params
    }

>     ccmID = sha256(encode(crossChainMessageSchema, ccm))

Reviewer: I would move this line further below (just before it is needed), in particular, so that we run ` validateFormat(ccm)` before calling encode here.

Reviewer 2: Sounds good. I thought that this was already fixed, but perhaps broken again at some point.

Reviewer: Fixed.

* Within each substore, check that all entries have a different `storeKey`.

> * For each entry in the chain data substore there must be a corresponding entry (with the same value of `storeKey`) in the outbox root, channel data, and chain validators substores and viceversa. However, if the chain account has status set to terminated, the entry in the outbox root must *not* be present.

Reviewer: This dependency somehow suggest that the chain data, channel data and chain validators substores could be merged into one.

Author: I'd prefer to keep it as it is.

Reviewer: Okay, all good.

* For each entry in the terminated outbox substore there must be a corresponding entry (with the same value of `storeKey`) in the terminated state substore.

> * For each entry in the terminated state substore, if the property `initialized` is set to `false`, the corresponding entry in the terminated outbox substore must *not* be present. Furthermore, the `stateRoot` must be set to empty bytes and `mainchainStateRoot` to a 32-bytes value. Conversely, if the property `initialized` is set to `True`, the `stateRoot` must be set to a 32-bytes value and `mainchainStateRoot` to empty bytes.

Reviewer: It should be `False` instead of `false`.

> * For each entry in the terminated state substore, if the property `initialized` is set to `false`, the corresponding entry in the terminated outbox substore must *not* be present. Furthermore, the `stateRoot` must be set to empty bytes and `mainchainStateRoot` to a 32-bytes value. Conversely, if the property `initialized` is set to `True`, the `stateRoot` must be set to a 32-bytes value and `mainchainStateRoot` to empty bytes.

Author: @gkoumout I used constants but I think it is fine if we repeat explicitly the checks here

* For each entry `chainValidators` in `interoperabilityAsset.chainValidatorsSubstore`, let `activeValidators = chainValidators.storeValue.activeValidators` and let `certificateThreshold = chainValidators.storeValue.certificateThreshold`:
  * `activeValidators` must have at least 1 element and at most `MAX_NUM_VALIDATORS` elements.

>   * `activeValidators` must be ordered lexicographically by `blsKey` property.

Reviewer: Regarding the following line below:

> Either `channel.messageFeeTokenID={chainID=1, localID=0}` (corresponding to the LSK token), or `channel.messageFeeTokenID.chainID` must be equal to `channelData.storeKey` or `ownChainAccount.ID` (the message fee token must be a native token of either chains).

Note that token IDs are now byte values and not objects so the notation should be adjusted according to the Token module.

Author: yes, fixed

  * Check that `⌊ ⅓ x totalWeight⌋ + 1 <= certificateThreshold <= totalWeight`, where `⌊⋅⌋` is the floor function.

> * If a chain account for another sidechain is present, than a chain account for the mainchain (`chainID = 1`) must be present, as well as the own chain account.

Reviewer: ```suggestion
* If a chain account for another sidechain is present, than a chain account for the mainchain (`chainID == CHAIN_ID_MAINCHAIN`) must be present, as well as the own chain account.
```

Author: fixed

            "fieldNumber": 1
          },
          "storeValue": {

>             ...terminatedStateAccountSchema,

Author: removed
